treat hermit like wasm32

m-mueller678 opened the pull request:
This implements stacker on hermit by treating it the same as wasm32

nagisa commented:
Could you please elaborate on why this implementation precisely is the right thing to do for the target and why e.g. `psm`-based implementation does not work? Is hermit running exclusively on esoteric architectures of some sort?

m-mueller678 commented:
I forgot to mention, this intends to fix #108. If I can get an implementation that just compiles and does nothing, I would be happy too.

I chose the wasm implementation simply because I am not familiar enough with psm to adapt the psm implementation to do the right thing for Hermit.

From what I could gather from the source code, they differ in how they allocate memory and how they query page size. Hermit supports neither mmap nor sysconf, so I went with the wasm implementation.

Hermit usually runs on x86_64 and aarch64, though there might also be other supported architectures.

nagisa commented on src/lib.rs:
```diff
@@ -258,9 +258,9 @@ psm_stack_manipulation! {
+            #[cfg(not(any(target_arch = "wasm32",target_os = "hermit")))]
             unsafe { libc::sysconf(libc::_SC_PAGE_SIZE) as usize }
-            #[cfg(target_arch = "wasm32")]
+            #[cfg(any(target_arch = "wasm32",target_os = "hermit"))]
```
If there is absolutely no way to obtain the page size on hermit then lets hardcode it to something more appropriate for platforms hermit runs on (probably 4kB.)

m-mueller678 replied:
Since this implementation does not use `mmap` to allocate, I don't see a good reason to allocate in multiple of the page size. Is page_size used for anything but controlling the size of allocations? If not, renaming to allocation_granularity or such might be appropriate.

Hermit uses Talc as an Allocator. [Here is how Talc rounds up allocation sizes](https://docs.rs/talc/latest/src/talc/talc.rs.html#327-374). As Talc stores some metadata along with each allocation, picking a power of two will have it just barely not fit within the size category. I could not find any special treatment for page size allocations, though maybe I missed it. Given this, an allocation granularity of 1 would make sense.

nagisa replied:
Yeah, I think its fine to call it `allocation_granularity` or perhaps even better -- to something with `alignment` in the name. E.g. stacks usually need to be at least 16 byte aligned on x64 regardless of the allocation strategy so it makes sense to have that as a minimum value on x64.

m-mueller678 replied:
Should `page_size` even be 2^16 for wasm, then? The wasm implementation later just allocates from the global allocator with alignment 16. I think it would make sense to move the page-number calculation into `StackRestoreGuard`, since it also doesn't make much sense for wasm and hermit.

m-mueller678 commented:
Is page_size used for anything but controlling the size of allocations? If not, renaming to allocation_granularity or such might be appropriate.

Hermit uses Talc as an Allocator. [Here is how Talc rounds up allocation sizes](https://docs.rs/talc/latest/src/talc/talc.rs.html#327-374). As Talc stores some metadata along with each allocation, picking a power of two will have it just barely not fit within the size category. I could not find any special treatment for page size allocations, though maybe I missed it. Given this, an allocation granularity of 1 would make sense.

m-mueller678 commented:
I moved `StackRestoreGuard` into its own module with two different implementations depending on target. I tried to leave the logic for mmap-using targets exactly the same. Some tings seemed wrong, I added to-dos for those.

nagisa commented on src/mmap_stack_restore_guard.rs:
```diff
@@ -0,0 +1,107 @@
+            .checked_mul(page_size)
+            .expect("unreasonably large stack requested");
+
+        // Next, there are a couple of approaches to how we allocate the new stack. If it is
+        // available, we take the most obvious path and use `mmap`.
```
```suggestion
```

This comment no longer makes much sense in the context.

nagisa commented on src/mmap_stack_restore_guard.rs:
```diff
@@ -0,0 +1,107 @@
+                old_stack_limit: get_stack_limit(),
+            };
+            // We leave a guard page without read/write access in our allocation.
+            // TODO we allocated two extra pages for guard pages, but here we only use one?
```
```suggestion
            // There is one guard page below the stack and another above it.
```

m-mueller678 replied:
Shouldn't it be `size_with_guard - 2 * page_size` here?
```rust
let result = libc::mprotect(
  above_guard_page,
  size_with_guard - page_size,
  libc::PROT_READ | libc::PROT_WRITE,
);
```

nagisa replied:
oh yeah, that's a good spot, wanna PR this?

m-mueller678 replied:
Yes, will open one later today

nagisa commented on src/lib.rs:
```diff
@@ -135,123 +135,30 @@ fn set_stack_limit(l: Option<usize>) {
+                // `StackRestoreGuard` allocates a memory area with suitable size and alignment.
+                // It also sets up stack guards if supported on target.
+                let guard = StackRestoreGuard::new(requested_stack_size);
+                let (stack_base,allocated_stack_size) = guard.stack_area();
```
```suggestion
                let (stack_base, allocated_stack_size) = guard.stack_area();
```

nagisa reviewed (approved):
Looks good to me, thank you!
